lib/elixir/test/elixir/module/types/types_test.exs

```diff
@@ -52,4 +105,388 @@ defmodule Module.Types.TypesTest do
+  end
+
+  describe "map warnings" do
+    test "show map() when comparing against non-map" do
+      string =
+        warning(
+          [foo],
+          (
+            foo.bar
+            :atom = foo
+          )
+        )
+
+      assert string == """
+             incompatible types:
+
+                 map() !~ :atom
+
+             in expression:
+
+                 # types_test.ex:4
+                 :atom = foo
+
+             where "foo" was given the type map() (due to calling var.field) in:
+
+                 # types_test.ex:3
+                 foo.bar
+
+             where "foo" was given the type :atom in:
+
+                 # types_test.ex:4
+                 :atom = foo
+
+             HINT: "var.field" (without parentheses) implies "var" is a map() while \
+             "var.fun()" (with parentheses) implies "var" is an atom()
+             """
+    end
+
+    test "use module as map (without parentheses)" do
+      string =
+        warning(
+          [foo],
+          (
+            %module{} = foo
+            module.__struct__
+          )
+        )
+
+      assert string == """
+             incompatible types:
+
+                 map() !~ atom()
+
+             in expression:
+
+                 # types_test.ex:4
+                 module.__struct__
+
+             where "module" was given the type atom() in:
+
+                 # types_test.ex:3
+                 %module{}
+
+             where "module" was given the type map() (due to calling var.field) in:
+
+                 # types_test.ex:4
+                 module.__struct__
+
+             HINT: "var.field" (without parentheses) implies "var" is a map() while \
+             "var.fun()" (with parentheses) implies "var" is an atom()
+             """
+    end
+
+    test "use map as module (with parentheses)" do
+      string = warning([foo], [is_map(foo)], foo.__struct__())
+
+      assert string == """
+             incompatible types:
+
+                 map() !~ atom()
+
+             in expression:
+
+                 # types_test.ex:1
+                 foo.__struct__()
+
+             where "foo" was given the type map() in:
+
+                 # types_test.ex:1
+                 is_map(foo)
+
+             where "foo" was given the type atom() (due to calling var.fun()) in:
+
+                 # types_test.ex:1
+                 foo.__struct__()
+
+             HINT: "var.field" (without parentheses) implies "var" is a map() while \
+             "var.fun()" (with parentheses) implies "var" is an atom()
+             """
+    end
+
+    test "non-existant map field warning" do
+      string =
+        warning(
+          (
+            map = %{foo: 1}
+            map.bar
+          )
+        )
+
+      assert string == """
+             undefined field "bar" in expression:
+
+                 # types_test.ex:3
+                 map.bar
+
+             expected one of the following fields: foo
+
+             where "map" was given the type map() in:
```

Comment:
How come we show `map()` instead of `%{foo: integer()}` here? Since it failed because of `map.bar` we should show the actual type instead of simplifying to `map()` which is a compatible type.